zcash_client_backend/src/data_api.rs

```diff
@@ -509,24 +536,34 @@ pub trait WalletWrite: WalletRead {
     /// The type of identifiers used to look up transparent UTXOs.
     type UtxoRef;
 
-    /// Tells the wallet to track the next available account-level spend authority, given
-    /// the current set of [ZIP 316] account identifiers known to the wallet database.
+    /// Tells the wallet to track the next available account-level spend authority, given the
+    /// current set of [ZIP 316] account identifiers known to the wallet database.
+    ///
+    /// Returns the account identifier for the newly-created wallet database entry, along with the
+    /// associated [`UnifiedSpendingKey`].
     ///
-    /// Returns the account identifier for the newly-created wallet database entry, along
-    /// with the associated [`UnifiedSpendingKey`].
+    /// If a birthday height is provided and that height is below the current chain tip, this
+    /// operation will trigger a re-scan of the blocks at and above the provided height; otherwise,
+    /// the account's birthday will be set to the current chain tip. The birthday height is defined
+    /// as the maximum block height that is known to contain no funds for the wallet. Ordinarily,
+    /// callers should set this to `chain_tip_height - PRUNING_DEPTH` for new wallets to ensure
+    /// that chain reorgs cannot cause loss of funds as a consequence of a transaction targeted at
+    /// a height greater than the chain tip instead being mined at a height below that chain tip as
+    /// part of a reorg.
```

Comment:
Hm, this is tricky because the `chain_tip_height - PRUNING_DEPTH` might be below Sapling activation. I guess that's okay, it just makes things weird.

Reply:
Actually, we need to make sure that we have the note commitment tree frontier when a wallet is created. Maybe make the birthday entirely non-optional?

Reply:
When `birthday = None`, the wallet historically set the birthday to "whatever height it has the most recent frontier for", rather than "chain tip".
- For the first account (previously added via `init_accounts_table`), that was the initial frontier (added via `init_blocks_table`).
- For subsequent accounts, it happened to be chain tip, because previously we used linear scanning so we'd always end up making the frontier at that height accessible.

Now that both of those methods are being replaced by `create_account`, and we aren't assuming linear scanning anymore, we need to ensure both cases are handled.
- For the first call to `create_account`, we want to have a frontier to bound the scanning work necessary.
  - If `birthday = Some(_)` then we're fine, as the caller provides a frontier.
  - If `birthday = None`, then the internally-selected birthday height will be missing a frontier, and we need to ensure that the scan queue is updated in a way that will produce it.
- For subsequent calls to `create_account`:
  - If `birthday = Some(_)` then we're fine, even if that birthday is earlier than any known to the wallet (as the provided frontier moves the bound on necessary scanning work a bit earlier, but doesn't otherwise break the assumption that we don't need to scan the tree earlier than the earlierst birthday height).
  - If `birthday = None` then we can just use the height we want (ensuring the scan queue will rescan after that height), and we don't need a frontier at all. We can instead rely on the fact that we will encounter a note for the account after this height, and all we actually need is a path at that note, not a frontier at the account's birthday height.

Reply:
After an offline discussion, we decided to make the birthday non-optional. This means that wallet developers can get predictable results, and we don't have to worry about a bunch of different cases here.